File: 02-GettingStarted/samples/java/McpCalculatorServer.java
// McpCalculatorServer.java - Sample MCP Calculator Server implementation in Java
package com.example.mcp;

import com.mcp.server.McpServer;
import com.mcp.tools.Tool;
import com.mcp.tools.ToolRequest;
import com.mcp.tools.ToolResponse;
import com.mcp.tools.ToolExecutionException;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class McpCalculatorServer {
    public static void main(String[] args) {
        // Create and configure MCP server
        McpServer server = new McpServer.Builder()
            .setName("Calculator MCP Server")
            .setVersion("1.0.0")
            .setPort(5000)
            .build();
        
        // Register calculator tool
        server.registerTool(new CalculatorTool());
        
        // Start server
        server.start();
        System.out.println("Calculator MCP Server started on port 5000");
    }
}

class CalculatorTool implements Tool {
    @Override
    public String getName() {
        return "calculator";
    }
    
    @Override
    public String getDescription() {
        return "Performs basic arithmetic operations";
    }
    
    @Override
    public Object getSchema() {
        // Define parameter schema
        Map<String, Object> schema = new HashMap<>();
        schema.put("type", "object");
        
        Map<String, Object> properties = new HashMap<>();
        
        Map<String, Object> operation = new HashMap<>();
        operation.put("type", "string");
        operation.put("enum", Arrays.asList("add", "subtract", "multiply", "divide"));
        properties.put("operation", operation);
        
        Map<String, Object> a = new HashMap<>();
        a.put("type", "number");
        properties.put("a", a);
        
        Map<String, Object> b = new HashMap<>();
        b.put("type", "number");
        properties.put("b", b);
        
        schema.put("properties", properties);
        schema.put("required", Arrays.asList("operation", "a", "b"));
        
        return schema;
    }
    
    @Override
    public ToolResponse execute(ToolRequest request) {
        // Extract parameters
        String operation = request.getParameters().get("operation").asText();
        double a = request.getParameters().get("a").asDouble();
        double b = request.getParameters().get("b").asDouble();
        
        double result = 0;
        
        // Perform calculation
        switch (operation) {
            case "add":
                result = a + b;
                break;
            case "subtract":
                result = a - b;
                break;
            case "multiply":
                result = a * b;
                break;
            case "divide":
                if (b == 0) {
                    throw new ToolExecutionException("Cannot divide by zero");
                }
                result = a / b;
                break;
            default:
                throw new ToolExecutionException("Unknown operation: " + operation);
        }
        
        // Create response
        Map<String, Object> response = new HashMap<>();
        response.put("result", result);
        
        return new ToolResponse.Builder()
            .setResult(response)
            .build();
    }
}

File: 02-GettingStarted/samples/python/mcp_calculator_server.py
#!/usr/bin/env python3
"""
Sample MCP Calculator Server implementation in Python.

This module demonstrates how to create a simple MCP server with a calculator tool
that can perform basic arithmetic operations (add, subtract, multiply, divide).
"""

import json
import logging
from typing import Dict, Any, Optional

from mcp_server import McpServer
from mcp_tools import Tool, ToolRequest, ToolResponse, ToolExecutionException


class CalculatorTool(Tool):
    """
    A tool that performs basic arithmetic operations.
    
    This tool handles add, subtract, multiply, and divide operations
    on two numerical operands.
    """
    
    def get_name(self) -> str:
        """Return the name of the tool."""
        return "calculator"
        
    def get_description(self) -> str:
        """Return a description of the tool."""
        return "Performs basic arithmetic operations"
    
    def get_schema(self) -> Dict[str, Any]:
        """
        Define the schema for the calculator tool.
        
        The schema specifies the required parameters and their types.
        
        Returns:
            Dict[str, Any]: The JSON schema definition
        """
        return {
            "type": "object",
            "properties": {
                "operation": {
                    "type": "string",
                    "enum": ["add", "subtract", "multiply", "divide"],
                    "description": "The arithmetic operation to perform"
                },
                "a": {"type": "number", "description": "First operand"},
                "b": {"type": "number", "description": "Second operand"}
            },
            "required": ["operation", "a", "b"]
        }
      def execute(self, request: ToolRequest) -> ToolResponse:
        """
        Execute the calculation based on the provided operation and operands.
        
        Args:
            request (ToolRequest): The request containing operation parameters
                                  
        Returns:
            ToolResponse: Response containing the calculation result
            
        Raises:
            ToolExecutionException: If division by zero is attempted or an unknown
                                   operation is provided
        """
        # Extract parameters
        params = request.parameters
        operation = params.get("operation")
        a = params.get("a")
        b = params.get("b")
        
        # Perform calculation
        if operation == "add":
            result = a + b
        elif operation == "subtract":
            result = a - b
        elif operation == "multiply":
            result = a * b
        elif operation == "divide":
            if b == 0:
                raise ToolExecutionException("Cannot divide by zero")
            result = a / b
        else:
            raise ToolExecutionException(f"Unknown operation: {operation}")
        
        # Return result
        return ToolResponse(
            result={"result": result}
        )


def configure_logging():
    """Configure basic logging for the application."""
    logging.basicConfig(
        level=logging.INFO,
        format="%(asctime)s - %(name)s - %(levelname)s - %(message)s",
        handlers=[logging.StreamHandler()]
    )


def main():
    """Initialize and start the MCP server with calculator tool."""
    # Configure logging
    configure_logging()
    logger = logging.getLogger(__name__)
    
    # Initialize server
    server = McpServer(
        name="Calculator MCP Server",
        version="1.0.0",
        port=5000
    )
    
    # Register tools
    calculator_tool = CalculatorTool()
    server.register_tool(calculator_tool)
    
    # Start server
    logger.info("Starting Calculator MCP Server on port 5000")
    server.start()


if __name__ == "__main__":
    main()
